File: src/systems/EconomyAlertAdapter.js
import {
  ALERT_DURATION_KINDS,
  ALERT_FOCUS_ACTIONS,
  ALERT_SEVERITIES,
  ALERT_TYPES
} from '../alerts/AlertService.js';
import { FISCAL_STATES } from './EconomyBalance.js';

/** Publishes one deduplicated, structured warning for authoritative fiscal stress. */
export class EconomyAlertAdapter {
  constructor({ economySystem, alertService } = {}) {
    if (!economySystem?.subscribe) throw new TypeError('economySystem must expose subscribe()');
    if (!alertService?.publish || !alertService?.resolve) {
      throw new TypeError('alertService must expose publish() and resolve()');
    }
    this.economy = economySystem;
    this.alerts = alertService;
    this.unsubscribe = economySystem.subscribe(event => this.sync(event.current), { emitCurrent: true });
  }

  sync(snapshot) {
    const fiscal = snapshot?.fiscal;
    if (!fiscal) return false;
    const key = 'economy:fiscal-recovery';
    if (fiscal.status === FISCAL_STATES.STABLE) {
      this.alerts.resolve(key, 'Recurring cashflow and emergency reserves are stable');
      return true;
    }
    const insolvent = fiscal.status === FISCAL_STATES.INSOLVENT;
    const recovering = fiscal.status === FISCAL_STATES.RECOVERY;
    this.alerts.publish({
      dedupeKey: key,
      type: ALERT_TYPES.ECONOMY,
      severity: insolvent ? ALERT_SEVERITIES.CRITICAL : ALERT_SEVERITIES.WARNING,
      title: insolvent
        ? 'City Capital exhausted — recovery available'
        : recovering ? 'Fiscal recovery restrictions active' : 'City budget is running a deficit',
      cause: `${fiscal.explanation} Net cashflow is ${snapshot.budgetBreakdown.netRate >= 0 ? '+' : '−'}$${Math.abs(Math.round(snapshot.budgetBreakdown.netRate * 60)).toLocaleString()}/min.`,
      location: 'Citywide treasury',
      duration: { kind: ALERT_DURATION_KINDS.UNTIL_RESOLVED },
      recommendation: fiscal.assistanceEligible
        ? `Claim the $${fiscal.emergencyGrant.toLocaleString()} stabilization grant, then complete contracts or remove optional upkeep.`
        : fiscal.actions.join(' '),
      relatedEntityIds: [],
      focusAction: { type: ALERT_FOCUS_ACTIONS.NONE }
    });
    return true;
  }

  dispose() {
    this.unsubscribe?.();
  }
}

export default EconomyAlertAdapter;

File: godot/MetroPulse.Domain/Alerts/EconomyAlertAdapter.cs
using System.Globalization;
using MetroPulse.Domain.Economy;

namespace MetroPulse.Domain.Alerts;

/// <summary>Publishes one deduplicated structured alert for authoritative fiscal stress.</summary>
public sealed class EconomyAlertAdapter : IDisposable
{
    private readonly AlertService alerts;
    private readonly Func<bool> unsubscribe;

    public EconomyAlertAdapter(EconomyLedger economy, AlertService alerts)
    {
        ArgumentNullException.ThrowIfNull(economy);
        this.alerts = alerts ?? throw new ArgumentNullException(nameof(alerts));
        unsubscribe = economy.Subscribe(economyEvent => Sync(economyEvent.Current), emitCurrent: true);
    }

    public bool Sync(EconomyLedgerSnapshot snapshot)
    {
        ArgumentNullException.ThrowIfNull(snapshot);
        const string key = "economy:fiscal-recovery";
        EconomyFiscalOverview fiscal = snapshot.Fiscal;
        if (fiscal.Status == FiscalStates.Stable)
        {
            alerts.Resolve(key, "Recurring cashflow and emergency reserves are stable");
            return true;
        }

        bool insolvent = fiscal.Status == FiscalStates.Insolvent;
        bool recovering = fiscal.Status == FiscalStates.Recovery;
        double perMinute = snapshot.BudgetBreakdown.NetRate * 60;
        string sign = perMinute >= 0 ? "+" : "−";
        string cashflow = JavascriptRound(Math.Abs(perMinute)).ToString("N0", CultureInfo.InvariantCulture);
        string recommendation = fiscal.AssistanceEligible
            ? $"Claim the ${fiscal.EmergencyGrant.ToString("N0", CultureInfo.InvariantCulture)} stabilization grant, then complete contracts or remove optional upkeep."
            : string.Join(' ', fiscal.Actions);

        alerts.Publish(new AlertInput
        {
            DedupeKey = key,
            Type = AlertTypes.Economy,
            Severity = insolvent ? AlertSeverities.Critical : AlertSeverities.Warning,
            Title = insolvent
                ? "City Capital exhausted — recovery available"
                : recovering ? "Fiscal recovery restrictions active" : "City budget is running a deficit",
            Cause = $"{fiscal.Explanation} Net cashflow is {sign}${cashflow}/min.",
            Location = AlertLocation.Named("Citywide treasury"),
            Duration = new AlertDuration { Kind = AlertDurationKinds.UntilResolved },
            Recommendation = recommendation,
            RelatedEntityIds = Array.Empty<string>(),
            FocusAction = new AlertFocusAction(),
        });
        return true;
    }

    public void Dispose() => unsubscribe();

    private static double JavascriptRound(double value) => Math.Floor(value + 0.5);
}
